adapters/oidc/js/src/main/resources/login-status-iframe.html

```diff
@@ -54,7 +54,11 @@
                         }
                         if (!cookie) {
                             if (sessionState != '') {
-                                callback('changed');
+                                // Adapter was authenticated (session ID exists) but session cookie is not found.
+                                // Either user was logged out shortly after authentication but before this iframe could be
+                                // loaded (usually just a few seconds), which is really a corner case, or 3rd part cookies
+                                // are blocked which is the most probable scenario.
+                                callback('cookie_not_found');
```

Comment:
You can't do this I'm afraid, as it is not correct behaviour with regards to the spec. According to the spec it should send unchanged if expected session matches actual state, or changed if it doesn't. See https://openid.net/specs/openid-connect-session-1_0.html#RPiframe

Reply:
@stianst I was thinking the same first, but then I realised this piece of code is visited only on the first iframe call. Every other call already has the `init` object/variable created so it will end up here: https://github.com/keycloak/keycloak/blob/master/adapters/oidc/js/src/main/resources/login-status-iframe.html#L71

I can't see any other way how to find out whether 3rd party cookies are blocked. WDYT?

Reply:
Thanks for pointing this out. jFTR, with @stianst we were discussing changing the callback here to `unchanged`. But this brings other problems.

The main problem is how to tell the adapter that iframe doesn't work. So as I see it, we've got following options.

**Option 1**
* Fully comply with specs, i.e. return `unchanged`.
* Adapter doesn't know that iframe is not working.
* Standard check-sso (the one without iframe, not the silent one) won't work:
https://github.com/keycloak/keycloak/blob/ec9bf6206e3d937a974ac84b9618c00823edc884/adapters/oidc/js/src/main/resources/keycloak.js#L283-L288
This will have to be reworked not to rely on iframe at all.A redirect to KC would always be performed even if the user is not logged in. Regardless if iframe is enabled or not because we can't know if it works or not.
* Initialization with token also won't work as before:
https://github.com/keycloak/keycloak/blob/ec9bf6206e3d937a974ac84b9618c00823edc884/adapters/oidc/js/src/main/resources/keycloak.js#L319-L331
The token won't be checked.
* Silent check-sso cannot work and user will always have to invoke login manually.

**Option 2**
* Hybrid solution: fully comply with the specs by default, but our JS adapter would use a special "mode" of the login iframe so that it would use `cookie_not_found` in the callback.
* Everything would work as before, except obviously the silent check-sso and single logout.

Reply:
Scratch that, option 2 won't work. In case of non-silent check-sso, it wouldn't work. Even if the iframe would return `cookie_not_found`, it wouldn't return it in case of check-sso – it would always return `unchaged` which would make check-sso completely unusable with 3rd party cookies blocked. At the same time, forcing the browser to redirect to KC with `prompt=none` even if cookies aren't blocked is unnecessary overhead, so it's not an option either.

I'll add an explicit and independent check for 3rd party cookies blocking when initializing the iframe, something like: https://github.com/mindmup/3rdpartycookiecheck